Before we do any work, we need to import several functions from cdapython:
- `Q` and `query` which power the search
- `columns` which lets us view entity field names
- `unique_terms` which lets view entity field contents

We're also asking cdapython to report it's version so we can be sure we're using the one we mean to.

In [1]:
from cdapython import Q, columns, unique_terms,query, constantVariables
import cdapython
print(cdapython.__file__)
print(cdapython.__version__)
integration_host = "http://35.192.60.10:8080/"
Q.set_host_url(integration_host)

/Users/amanda.charbonneau/github/cda-python/cdapython/__init__.py
2022.5.23


CDA data comes from three sources:
- The [Proteomic Data Commons](https://proteomic.datacommons.cancer.gov/pdc/) (PDC)
- The [Genomic Data Commons](https://gdc.cancer.gov/) (GDC)
- The [Imaging Data Commons](https://datacommons.cancer.gov/repository/imaging-data-commons) (IDC)

The CDA makes this data searchable in four main endpoints:

- `subject`: Specific, unique, individuals
- `researchsubject`: Study-individual aggregate entities. A Subject who was part of three studies will appear as three ResearchSubjects
- `specimen`: Samples taken from individual
- `file`: Data about Subjects, ResearchSubjects, Specimens, and their associated information

and two endpoints that offer deeper information about data in the researchsubject endpoint:

- `diagnosis`: Information about what medical diagnosis a researchsubject has
- `treatment`: Information about what medical treatment(s) were performed for a given diagnosis

If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specfic format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.

In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.

In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. 



## Basic search with endpoints

Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`:

In [2]:
myquery = Q('ResearchSubject.primary_diagnosis_site = "brain"')


<div style="background-color:#6ce6b9;color:black;padding:20px;">
<h3>Where did those terms come from?</h3>
    
If you aren't sure how we knew what terms to put in our search, please refer back to the <a href="../SearchTerms">What search terms are available?</a> notebook. 
</div>

### subject
Now we can use that query to search any of information types. Let's start by looking at what Subjects meet our criteria. To do that, we will send our query to the subject endpoint, then ask for it to run:

In [3]:
subjectresults = myquery.subject.run()

Getting results from database

Total execution time: 3748 ms


We saved the output in a variable `subjectresults`, so we don't get much visible output. To see what our results are, we need to look into the variable. The simplest way is to call `subjectresults` directly:

In [4]:
subjectresults


            QueryID: 6b1f5113-18d4-40f0-9a49-4062a5653fea
            
            Offset: 0
            Count: 100
            Total Row Count: 2314
            More pages: True
            

This output tells us our QueryID, which we don't really need, but the computer does to track our questions. Then it tells us four parameters that describe our results:

---

- **Offset:** This is how many rows of information we've told the query to skip in the data, here we didn't tell it to skip anything, so the offset is zero
- **Count:** This is how many rows the current page of our results table has. To keep searches fast, we default to pages with 100 rows.
- **Total Row Count:** This is how many rows are in the full results table
- **More pages:** This is always a True or False. False means that our current page has all the available results. True means that we will see only the first 100 results in this table, and will need to page through for more.

---
    
Now that we've seen the metadata about our results, let's look at the actual table. The easiest way to do this is by using the python function `.to_dataframe()` on our `subjectresults` variable:

In [5]:
subjectresults.to_dataframe()

id  \
0        ACRIN-FMISO-Brain-010   
1                      C140343   
2                    C3L-01327   
3                      C808110   
4          GENIE-GRCC-e6a259ca   
..                         ...   
95               QIN-GBM-TR-01   
96  RIDER Neuro MRI-2579032163   
97                TCGA-02-0336   
98                TCGA-02-0446   
99                TCGA-06-0397   

                                           identifier       species     sex  \
0   [{'system': 'IDC', 'value': 'ACRIN-FMISO-Brain...  Homo sapiens    None   
1             [{'system': 'PDC', 'value': 'C140343'}]  Homo sapiens    male   
2   [{'system': 'GDC', 'value': 'C3L-01327'}, {'sy...  Homo sapiens    male   
3             [{'system': 'PDC', 'value': 'C808110'}]  Homo sapiens  female   
4   [{'system': 'GDC', 'value': 'GENIE-GRCC-e6a259...  Homo sapiens  female   
..                                                ...           ...     ...   
95      [{'system': 'IDC', 'value': 'QIN-GBM-TR-01'}]  Homo sapiens    None   
96  [{'system': 'IDC', 'value': 'RIDER Neuro MRI-2...  Homo sapiens    None   
97       [{'system': 'GDC', 'value': 'TCGA-02-0336'}]  Homo sapiens    None   
98       [{'system': 'GDC', 'value': 'TCGA-02-0446'}]  Homo sapiens    male   
99       [{'system': 'GDC', 'value': 'TCGA-06-0397'}]  Homo sapiens  female   

                      race               ethnicity days_to_birth  \
0                     None                    None          None   
1                    white  not hispanic or latino          None   
2                    white  not hispanic or latino        -27160   
3                    white  not hispanic or latino          None   
4   not allowed to collect  not allowed to collect        -17166   
..                     ...                     ...           ...   
95                    None                    None          None   
96                    None                    None          None   
97                    None                    None          None   
98                   white  not hispanic or latino        -22541   
99                   white  not hispanic or latino        -20845   

                           subject_associated_project  vital_status  \
0                                 [acrin_fmiso_brain]          None   
1   [Proteogenomic Analysis of Pediatric Brain Can...         Alive   
2              [cptac_gbm, CPTAC3-Discovery, CPTAC-3]          Dead   
3   [Proteogenomic Analysis of Pediatric Brain Can...         Alive   
4                                        [GENIE-GRCC]  Not Reported   
..                                                ...           ...   
95                       [qin_gbm_treatment_response]          None   
96                                  [rider_neuro_mri]          None   
97                                         [TCGA-GBM]          None   
98                                         [TCGA-GBM]          Dead   
99                                         [TCGA-GBM]          Dead   

   age_at_death  cause_of_death  
0          None            None  
1          None    Not Reported  
2           646  Cancer Related  
3          None    Not Reported  
4          None            None  
..          ...             ...  
95         None            None  
96         None            None  
97         None            None  
98          282            None  
99          274            None  

[100 rows x 11 columns]

By default `to_dataframe()` shows us the first and last five rows for the first page of our results, so we can easily preview our data.

Since we queried the Subject endpoint, our default results tell us Subject level information, that is, information about unique individuals: their sex, race, age, species, etc. The `id` column tells us the unique identifier for each individual. The identifier column has nested information about what study or studies a Subject participated in, and will list all of their researchsubject identifiers. 

The `to_dataframe()` function converts the results to a pandas dataframe. So if we save the dataframe to a variable, we can use any pandas functions to explore it. For example, lets see whether any of the Subjects in our first 100 results are black or african american. First we'll save the results to a dataframe, then subset that dataframe to only show rows where the word "black" appears in the "race" column. "NAs" which are shown as "None" in these tables, so for our filter to work, we'll need to specifically tell it to ignore NAs. We're also telling it we want the word "black" regardless of capitalization with `case=False`:


In [6]:
subjectdata = subjectresults.to_dataframe()
subjectdata[subjectdata['race'].str.contains("black", case=False, na=False)]

id                                    identifier       species  \
38  TCGA-06-A7TL  [{'system': 'GDC', 'value': 'TCGA-06-A7TL'}]  Homo sapiens   
69  TCGA-26-A7UX  [{'system': 'GDC', 'value': 'TCGA-26-A7UX'}]  Homo sapiens   

       sex                       race               ethnicity days_to_birth  \
38  female  black or african american  not hispanic or latino        -11297   
69    male  black or african american  not hispanic or latino        -17365   

   subject_associated_project vital_status age_at_death cause_of_death  
38                 [TCGA-GBM]        Alive         None           None  
69                 [TCGA-GBM]        Alive         None           None

There are three subjects in our first hundred results that meet the criteria. If we just want to be sure that the data contains some value, this might be good enough. But often we want to search the entire set of results and not just the first page. 

We'll cover how to work with large results dataframes in the [Pagination](../Pagination.ipynb) notebook. Or, learn how to get summary information from search results in the [Data Summaries](../DataSummaries.ipynb) notebook.


---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>Subject Field Definitions</h3>

<i>A subject is a specific, unique individual: for e.g. a single human. When consent allows, a given entity will have a single subject ID that can be connected to all their studies and data across all datasets</i>

    
<ul>
  <li><b>id:</b> The unique identifier for this subject</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the subject had there</li>
  <li><b>species:</b> The species of the subject</li>
  <li><b>sex:</b> A reference to the biological sex of the donor organism. </li>
  <li><b>race:</b> The race of the subject</li>
  <li><b>ethnicity:</b> The ethnicity of the subject</li>
  <li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days</li>
  <li><b>subject_associated_project:</b> An embedded array of the names of projects (studies) the subject was part of</li>
  <li><b>vital_status:</b> Whether the subject is alive</li>
  <li><b>age_at_death:</b> The number of days after first enrollment that the subject died</li>
  <li><b>cause_of_death:</b> The cause of death, if known</li>
</ul>  

</div>
    
---

### researchsubject

If we're interested in what researchsubjects meet our critera, we can also run our query against the researchsubject endpoint:

In [7]:
researchsubjectresults = myquery.researchsubject.run()
researchsubjectresults

Getting results from database

Total execution time: 3490 ms



            QueryID: 35ba8bad-9d0d-4fb5-9b9d-c2445caf707d
            
            Offset: 0
            Count: 100
            Total Row Count: 2923
            More pages: True
            

Now we see that our 2314 subjects have 2923 researchsubjects between them, that means that some, but not all, of our subjects were participants in more than one study. Let's peek at the data:

In [8]:
researchsubjectresults.to_dataframe()

id  \
0         104c3a87-2139-11ea-aee1-0e1aae319e49   
1         153c2442-ea61-4b5e-8c5c-c71d287c6055   
2         1edfe67c-1142-4498-927c-9b7b426af7a3   
3         3d58ded1-b7cf-420c-ba59-5b424683f0df   
4         559a8f6f-dd60-43b7-ae09-16cc4206064c   
..                                         ...   
95        723cad09-aa2d-4492-b19d-661b310e1c58   
96        7338e476-10ff-444a-a16f-429d28355f65   
97        834e560a-a506-456d-baf2-a206ca058153   
98  ACRIN-DSC-MR-Brain-037__acrin_dsc_mr_brain   
99                        C3L-01327__cptac_gbm   

                                           identifier  \
0   [{'system': 'PDC', 'value': '104c3a87-2139-11e...   
1   [{'system': 'GDC', 'value': '153c2442-ea61-4b5...   
2   [{'system': 'GDC', 'value': '1edfe67c-1142-449...   
3   [{'system': 'GDC', 'value': '3d58ded1-b7cf-420...   
4   [{'system': 'GDC', 'value': '559a8f6f-dd60-43b...   
..                                                ...   
95  [{'system': 'GDC', 'value': '723cad09-aa2d-449...   
96  [{'system': 'GDC', 'value': '7338e476-10ff-444...   
97  [{'system': 'GDC', 'value': '834e560a-a506-456...   
98  [{'system': 'IDC', 'value': 'ACRIN-DSC-MR-Brai...   
99          [{'system': 'IDC', 'value': 'C3L-01327'}]   

   member_of_research_project primary_diagnosis_condition  \
0            CPTAC3-Discovery                Glioblastoma   
1                    TCGA-GBM                     Gliomas   
2                    TCGA-GBM                     Gliomas   
3                    TCGA-LGG                     Gliomas   
4                    TCGA-LGG                     Gliomas   
..                        ...                         ...   
95                   TCGA-LGG                     Gliomas   
96                   TCGA-GBM                     Gliomas   
97                   TCGA-LGG                     Gliomas   
98         acrin_dsc_mr_brain                        None   
99                  cptac_gbm                        None   

   primary_diagnosis_site              subject_id  
0                   Brain               C3L-03407  
1                   Brain            TCGA-28-5209  
2                   Brain            TCGA-02-0339  
3                   Brain            TCGA-HT-7686  
4                   Brain            TCGA-HT-A74L  
..                    ...                     ...  
95                  Brain            TCGA-HT-A74O  
96                  Brain            TCGA-28-5213  
97                  Brain            TCGA-DU-7010  
98                  Brain  ACRIN-DSC-MR-Brain-037  
99                  Brain               C3L-01327  

[100 rows x 6 columns]

Each row from the researchsubject endpoint results tells us about a subject in a given study. Using this endpoint we can find out information like what studies fit our search criteria, and also get data that we can filter to have only subjects from multiple studies, or only subjects from single studies.

Any given subject will have one row per study they participated in. The subject_id in the last column of this view is the same as the `id` in the first column of the Subjects endpoint results. You can use this to combine information across endpoints, which is covered in the [Merging Results](../MergingResults.ipynb) notebook.


---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>ResearchSubject Field Definitions</h3>

<i>A research subject is the entity of interest in a research study, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subject’s privacy. An individual who participates in 3 studies will have 3 researchsubject IDs</i>
    
<ul>
  <li><b>id:</b> The unique identifier for this researchsubject</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the researchsubject had there</li>
  <li><b>member_of_research_project:</b> The name of the study/project that the subject particpated in</li>
  <li><b>primary_diagnosis_condition:</b> The cancer, disease or other condition under study</li>
  <li><b>primary_diagnosis_site:</b> The primary_disease_site that qualifies the researchsubject for the research_project</li>
  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>
</ul>  

</div>
    
---

### diagnosis

The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria:

In [9]:
diagnosisresults = myquery.diagnosis.run()
diagnosisresults.to_dataframe()

Getting results from database

Total execution time: 3452 ms


id  \
0   02ce9b1f-18f4-45ec-a487-63f9c74bb422   
1   0f7bb10c-a17e-4a98-b559-461edbf840c8   
2   13c464b8-bd62-5d47-91ff-4b2c9183c48b   
3   195c5853-a955-551b-8601-6f0d8a08cea8   
4   2faa8b42-f38e-50f2-9920-cafeafc989ae   
..                                   ...   
95  7a3112c5-a878-5b23-bdbc-3aa91d2cede6   
96  7b8b418c-70eb-5b2c-a880-1c869f324f21   
97  8e43bc6d-48f0-5567-8333-64c91a50d4eb   
98  91404a49-c46a-43c3-8bdc-6ef7e4feba82   
99  b82d770d-caf9-5420-bcc1-3ab57dbdaf9c   

                                           identifier  \
0   [{'system': 'GDC', 'value': '02ce9b1f-18f4-45e...   
1   [{'system': 'GDC', 'value': '0f7bb10c-a17e-4a9...   
2   [{'system': 'GDC', 'value': '13c464b8-bd62-5d4...   
3   [{'system': 'GDC', 'value': '195c5853-a955-551...   
4   [{'system': 'GDC', 'value': '2faa8b42-f38e-50f...   
..                                                ...   
95  [{'system': 'GDC', 'value': '7a3112c5-a878-5b2...   
96  [{'system': 'GDC', 'value': '7b8b418c-70eb-5b2...   
97  [{'system': 'GDC', 'value': '8e43bc6d-48f0-556...   
98  [{'system': 'GDC', 'value': '91404a49-c46a-43c...   
99  [{'system': 'GDC', 'value': 'b82d770d-caf9-542...   

          primary_diagnosis age_at_diagnosis morphology stage         grade  \
0     Mixed germ cell tumor             None     9085/3  None  Not Reported   
1              Glioblastoma            23121     9440/3  None  Not Reported   
2    Oligodendroglioma, NOS            12764     9450/3  None  not reported   
3              Glioblastoma            14838     9440/3  None  not reported   
4              Glioblastoma            28696     9440/3  None  not reported   
..                      ...              ...        ...   ...           ...   
95             Mixed glioma            20128     9382/3  None  not reported   
96             Glioblastoma            22332     9440/3  None  not reported   
97  Astrocytoma, anaplastic            16300     9401/3  None  not reported   
98             Glioblastoma            25423     9440/3  None  Not Reported   
99             Mixed glioma            12853     9382/3  None  not reported   

   method_of_diagnosis           subject_id  \
0                 None  GENIE-MSK-P-0002849   
1                 None            C3N-01366   
2                 None         TCGA-DU-5870   
3                 None         TCGA-08-0351   
4                 None         TCGA-02-0281   
..                 ...                  ...   
95                None         TCGA-P5-A77X   
96                None         TCGA-06-1800   
97                None         TCGA-P5-A736   
98                None            C3N-01857   
99                None         TCGA-S9-A7QY   

                      researchsubject_id  
0   3ecb27b2-75cc-44d0-879e-035e19f39193  
1   a1da2f64-5a24-4e6b-abc1-e70aae23082c  
2   eb42625e-fbbb-4bee-bbe9-315b3ca2138a  
3   91063e7e-60db-47a5-aeca-af7c37b1074e  
4   42fbe95f-9068-45a8-a8a2-c24db757eae6  
..                                   ...  
95  878ced76-3144-4141-896f-50e3e863df36  
96  b1ce95bf-7441-4aac-8131-19a177d2b8ac  
97  ac3779f6-cc8b-428b-9531-03562a614515  
98  9b84c1a2-d24d-4085-b597-eff3d2203756  
99  94a2ff09-f7ac-4aea-a2b2-0dc1b82ae635  

[100 rows x 10 columns]


---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>Diagnosis Field Definitions</h3>

<i>A diagnosis is a medical classification of a disease for a given research subject in a given study. A single research subject may have different diagnoses across different studies</i>

    
<ul>
  <li><b>id:</b> The unique identifier for this diagnosis in this research subject</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the diagnosed researchsubject had there</li>
  <li><b>primary_diagnosis:</b> The main medical diagnosis for this subject in this study</li>
  <li><b>age_at_diagnosis:</b> The subjects age in days after birth on the day they were first diagnosed</li>
  <li><b>morphology:</b> The <a href="https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology">International Classification of Diseases for Oncology</a> diagnosic code for this diagnosis</li>
  <li><b>stage:</b> A measure of disease spread. Different diseases may use different staging criteria, please refer to the originating data source to see what staging system is reported</li>
  <li><b>grade:</b> A measure of cell abnormality. Different diseases may use different grading criteria, please refer to the originating data source to see what grading system is reported</li>
  <li><b>method_of_diagnosis:</b> The test or system used for determining the diagnosis</li>
  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>
  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>
</ul>  

</div>
    
---




### treatment

The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:

In [10]:
treatmentresults = myquery.treatment.run()
treatmentresults.to_dataframe()

Getting results from database

Total execution time: 3535 ms


id  \
0   0611032f-6cde-5d42-b305-23219b4b486a   
1   08cabbfd-c1c2-5689-a1e6-ae03f0e8ecc9   
2   1e45d588-c532-5952-b8e5-56cc67267853   
3   2412b555-02be-5e02-8c26-cc4db13cb11c   
4   2afa81c1-efba-5fcd-b171-74820ac8d991   
..                                   ...   
95  34b6e7f0-6cca-5054-90a0-fb65073ebca9   
96  3c9a6945-d2a1-526c-aab5-edd8ecf3152e   
97  4440d709-1beb-5233-9781-ae767e91b0de   
98  48f9d02d-5dfe-5124-85d9-ec9fdd7d8f77   
99  4e7e5d91-f4fa-5ee1-89cf-45578413b59a   

                                           identifier  \
0   [{'system': 'GDC', 'value': '0611032f-6cde-5d4...   
1   [{'system': 'GDC', 'value': '08cabbfd-c1c2-568...   
2   [{'system': 'GDC', 'value': '1e45d588-c532-595...   
3   [{'system': 'GDC', 'value': '2412b555-02be-5e0...   
4   [{'system': 'GDC', 'value': '2afa81c1-efba-5fc...   
..                                                ...   
95  [{'system': 'GDC', 'value': '34b6e7f0-6cca-505...   
96  [{'system': 'GDC', 'value': '3c9a6945-d2a1-526...   
97  [{'system': 'GDC', 'value': '4440d709-1beb-523...   
98  [{'system': 'GDC', 'value': '48f9d02d-5dfe-512...   
99  [{'system': 'GDC', 'value': '4e7e5d91-f4fa-5ee...   

                 treatment_type treatment_outcome days_to_treatment_start  \
0        Radiation Therapy, NOS              None                    None   
1   Pharmaceutical Therapy, NOS              None                    None   
2   Pharmaceutical Therapy, NOS              None                    None   
3   Pharmaceutical Therapy, NOS              None                    None   
4        Radiation Therapy, NOS              None                    None   
..                          ...               ...                     ...   
95       Radiation Therapy, NOS              None                    None   
96       Radiation Therapy, NOS              None                    None   
97  Pharmaceutical Therapy, NOS              None                    None   
98       Radiation Therapy, NOS              None                    None   
99       Radiation Therapy, NOS              None                    None   

   days_to_treatment_end therapeutic_agent treatment_anatomic_site  \
0                   None              None                    None   
1                   None              None                    None   
2                   None              None                    None   
3                   None              None                    None   
4                   None              None                    None   
..                   ...               ...                     ...   
95                  None              None                    None   
96                  None              None                    None   
97                  None              None                    None   
98                  None              None                    None   
99                  None              None                    None   

   treatment_effect treatment_end_reason number_of_cycles    subject_id  \
0              None                 None             None  TCGA-08-0346   
1              None                 None             None  TCGA-RY-A845   
2              None                 None             None  TCGA-DU-8158   
3              None                 None             None  TCGA-HT-7477   
4              None                 None             None  TCGA-32-2615   
..              ...                  ...              ...           ...   
95             None                 None             None  TCGA-WH-A86K   
96             None                 None             None  TCGA-HT-7879   
97             None                 None             None  TCGA-12-1598   
98             None                 None             None  TCGA-06-1084   
99             None                 None             None  TCGA-E1-A7YU   

                      researchsubject_id          researchsubject_diagnosis_id  
0   5536201c-e739-46e6-8200-0bdcc28ae9ef  202921e5-a4e8-5d23-9d3c-de53394f93


---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>Treatment Field Definitions</h3>

<i><i>A treatment is a medical intervention for a diagnosed disease in a given subject in a given study. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>

    
<ul>
  <li><b>id:</b> The unique identifier for this treatment of this diagnosis in this research subject</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the treated researchsubject had there</li>
  <li><b>treatment_type:</b> The medical intervention undertaken</li>
  <li><b>treatment_outcome:</b> The result of the medical intervention</li>
  <li><b>days_to_treatment_start:</b> </li>
  <li><b>days_to_treatment_end:</b> </li>
  <li><b>therapeutic_agent:</b> What treatment or drug was used for this researchsubject</li>
  <li><b>treatment_anatomic_site:</b> The specific body location of the treatment</li>
  <li><b>treatment_effect:</b> </li>
  <li><b>treatment_end_reason:</b> </li>
  <li><b>number_of_cycles:</b> </li>
  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>
  <li><b>researchsubject_id:</b> An identifier for the researchsubject. Can be joined to the `id` field from researchsubject results</li>
  <li><b>researchsubject_diagnosis_id:</b> An identifier for the diagnosis. Can be joined to the `id` field from diagnosis results</li>
</ul>  

</div>
    
---




### specimens

We can use this same query to see what specimens are available for brain tissue at the CDA:

In [11]:
specresults =  myquery.specimen.run()
print(specresults)

Getting results from database

Total execution time: 3511 ms

            QueryID: e348a333-9d2a-419c-99b4-bb4fd44b7ce2
            
            Offset: 0
            Count: 100
            Total Row Count: 39150
            More pages: True
            


Nearly 40,000 specimens meet our search criteria! We would typically expect this number to be much larger than our number of subjects or researchsubjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. Since we didn't specify any further filters, our results will return all of these as seperate speciments. Let's look at a few:

In [12]:
specresults.to_dataframe()

id  \
0   018966a4-712b-4921-a7e7-02e390f43288   
1   0190d8c0-20ab-5475-8804-c99979e77dbe   
2   01ed9fd7-7d44-401b-a6b7-4fc6922813fd   
3   02117459-65da-4a3c-ba9a-ead2c0daa5ba   
4   023d5905-9dbf-40a7-90c1-761d3f48ad9a   
..                                   ...   
95  485cb0f3-e668-47a5-8b8a-0f8f81688ef2   
96  49280503-5136-4c10-8e42-061abbb6fa05   
97  49289dc2-31a6-427d-a7b9-e8e3613e17c1   
98  49ae4c42-847f-4da6-9a41-b9eaaa8831f1   
99  4a02dbff-6707-4e83-9d79-efe01846f763   

                                           identifier associated_project  \
0   [{'system': 'GDC', 'value': '018966a4-712b-492...           TCGA-GBM   
1   [{'system': 'GDC', 'value': '0190d8c0-20ab-547...           TCGA-LGG   
2   [{'system': 'GDC', 'value': '01ed9fd7-7d44-401...           TCGA-GBM   
3   [{'system': 'GDC', 'value': '02117459-65da-4a3...           TCGA-LGG   
4   [{'system': 'GDC', 'value': '023d5905-9dbf-40a...           TCGA-GBM   
..                                                ...                ...   
95  [{'system': 'GDC', 'value': '485cb0f3-e668-47a...           TCGA-GBM   
96  [{'system': 'GDC', 'value': '49280503-5136-4c1...           TCGA-GBM   
97  [{'system': 'GDC', 'value': '49289dc2-31a6-427...           TCGA-LGG   
98  [{'system': 'GDC', 'value': '49ae4c42-847f-4da...           TCGA-LGG   
99  [{'system': 'GDC', 'value': '4a02dbff-6707-4e8...           TCGA-GBM   

   age_at_collection primary_disease_type anatomical_site  \
0             -14729              Gliomas            None   
1             -14716              Gliomas            None   
2             -26247              Gliomas            None   
3             -17501              Gliomas            None   
4             -22417              Gliomas            None   
..               ...                  ...             ...   
95            -21768              Gliomas            None   
96            -27111              Gliomas            None   
97            -16280              Gliomas            None   
98            -17432              Gliomas            None   
99            -18195              Gliomas            None   

    source_material_type specimen_type                 derived_from_specimen  \
0   Blood Derived Normal       analyte  790f03cf-ad8f-4cde-8ba5-f465837c687f   
1          Primary Tumor       portion  5665435a-9b77-4504-a60a-b6d3edeb341f   
2          Primary Tumor         slide  37af7534-d545-45ee-9de3-387ad85d98f7   
3   Blood Derived Normal       portion  fa252999-283a-4ddc-84b5-17257939cc73   
4   Blood Derived Normal       analyte  19496c06-447b-437a-bae8-20d1b0c411c5   
..                   ...           ...                                   ...   
95         Primary Tumor       aliquot  6b524320-94c8-4035-b1d1-e1d3d3fe8679   
96         Primary Tumor       aliquot  3b705faf-6584-4dd8-b5f5-423ec3ee7e60   
97         Primary Tumor       aliquot  d1469517-114e-41da-9b50-ccecea62826b   
98  Blood Derived Normal       aliquot  c55220e7-4667-42a2-a8de-4641a29cae1d   
99         Primary Tumor       aliquot  6a6dc2fe-41ed-4d11-9ead-a2c1a1430f5b   

      subject_id                    researchsubject_id  
0   TCGA-06-0139  5bbeb8a7-3ac2-4ef1-bd48-6f83bc994a75  
1   TCGA-R8-A6MK  65836b59-b469-4b89-adf5-e65735e494f5  
2   TCGA-06-0644  652a39f0-c539-4303-bedc-f0d7f4669d9f  
3   TCGA-RY-A840  738143b3-2d82-42f0-9fd4-8cbabc9c9b4f  
4   TCGA-02-0046  5b8f05fa-0145-4142-8d41-5980e4d36d81  
..           ...                                   ...  
95  TCGA-12-5301  10efeb3c-ef55-4227-91b9-8c2bb4ff43d5  
96  TCGA-26-5134  20bad0d5-3135-49a4-a119-fa7a1e56fd1b  
97  TCGA-HT-7884  86aa5790-9afd-4653-926a-625516df1f56  
98  TCGA-EZ-7264  fc3b7596-f515-446f-81db-fed0154ca2c5  
99  TCGA-16-1045  cd73f38a-f6ff-4e7b-bd43-ec7b82dbe7df  

[100 rows x 11 columns]


---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>Specimen Field Definitions</h3>

<i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>

    
<ul>
  <li><b>id:</b> The unique identifier for this specimen</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the specimen had there</li>
  <li><b>associated_project:</b> The name of the study/project that the subject particpated in</li>
  <li><b>age_at_collection:</b> The subjects age in days (counting backwards to birth) on the day of the collection of the proximate specimen</li>
  <li><b>primary_disease_type:</b> The disease that qualifies the researchsubject for the associated_project</li>
  <li><b>anatomical_site:</b> The body part from which the proximate specimen was taken</li>
  <li><b>source_material_type:</b> The general kind of material from which the specimen was derived, indicating the physical nature of the source material</li>
  <li><b>specimen_type:</b> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide</li>
  <li><b>derived_from_specimen:</b> For derived samples, the `id` for the original sample</li>
  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>
  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>
</ul>  

</div>
    
---


### file

The file endpoint returns information about files that meet our search criteria, regardless of whether they are attached to subjects, research-subjects or specimens: 

In [13]:
myquery.file.run()

Getting results from database

Total execution time: 3522 ms



            QueryID: 1aa12ab0-ffe0-4588-8514-51c3960d8d5c
            
            Offset: 0
            Count: 100
            Total Row Count: 4530800
            More pages: True
            

In [14]:
fileresults = myquery.file.run()
fileresults.to_dataframe()

Getting results from database

Total execution time: 3805 ms


id  \
0   1b450a73-a16d-4de4-b140-9295335063ad   
1   ef900de0-892c-4e2e-8eae-74caf55de54c   
2   d43ef537-6cd0-446a-9bf0-fe0767ed6bcf   
3   58581feb-03c1-496c-9328-e8dda42c0d59   
4   1fd6817e-fa11-4569-a740-86584eaff394   
..                                   ...   
95  70b44166-c646-43eb-8d76-b567bf9328ac   
96  da7f4908-e98d-4ce6-93e3-a721276f733b   
97  8341914b-205d-4765-91ad-3e223c465459   
98  ade18349-cd37-4da0-94c2-2a9e664002ac   
99  e0c450dd-04f1-48e9-b572-19606de73d5b   

                                           identifier  \
0   [{'system': 'PDC', 'value': '1b450a73-a16d-4de...   
1   [{'system': 'GDC', 'value': 'ef900de0-892c-4e2...   
2   [{'system': 'GDC', 'value': 'd43ef537-6cd0-446...   
3   [{'system': 'GDC', 'value': '58581feb-03c1-496...   
4   [{'system': 'PDC', 'value': '1fd6817e-fa11-456...   
..                                                ...   
95  [{'system': 'PDC', 'value': '70b44166-c646-43e...   
96  [{'system': 'GDC', 'value': 'da7f4908-e98d-4ce...   
97  [{'system': 'PDC', 'value': '8341914b-205d-476...   
98  [{'system': 'GDC', 'value': 'ade18349-cd37-4da...   
99  [{'system': 'GDC', 'value': 'e0c450dd-04f1-48e...   

                                                label  \
0                         NCI-11Plex-2-F12-f10321.psm   
1   TCGA-LGG.c6bdb34c-fa93-4736-a842-afec0d9fa990....   
2   TCGA-R8-A6YH-01A-02-TS2.AD2BC93E-924F-46FC-A41...   
3   5c6ea055-2352-4111-9f8d-6eec3b9b3847.methylati...   
4            10CPTAC_GBM_W_PNNL_20190501_B3S2_f09.psm   
..                                                ...   
95                   NCI-11pex-6C-Phos-c05377.mzML.gz   
96  179f277a-8e72-4008-8a27-ea6e0d4ec7db.wxs.aliqu...   
97       10CPTAC_GBM_W_PNNL_20190501_B3S2_f24.mzid.gz   
98  962fbeae-2f42-4969-b35d-b1ec852c57ca.wxs.aliqu...   
99  6998fda2-aab1-4175-8606-2df81e5e7114_wxs_gdc_r...   

                  data_category                    data_type file_format  \
0      Peptide Spectral Matches                         Text         tsv   
1          Structural Variation            Transcript Fusion         TSV   
2                   Biospecimen                  Slide Image         SVS   
3               DNA Methylation       Methylation Beta Value         TXT   
4      Peptide Spectral Matches                         Text         tsv   
..                          ...                          ...         ...   
95       Processed Mass Spectra                Open Standard        mzML   
96  Simple Nucleotide Variation  Aggregated Somatic Mutation         MAF   
97     Peptide Spectral Matches                Open Standard   mzIdentML   
98  Simple Nucleotide Variation      Masked Somatic Mutation         MAF   
99             Sequencing Reads                Aligned Reads         BAM   

                                   associated_project  \
0   Proteogenomic Analysis of Pediatric Brain Canc...   
1                                            TCGA-LGG   
2                                            TCGA-LGG   
3                                             CPTAC-3   
4                                    CPTAC3-Discovery   
..                                                ...   
95  Proteogenomic Analysis of Pediatric Brain Canc...   
96                                           TCGA-GBM   
97                                   CPTAC3-Discovery   
98                                           TCGA-GBM   
99                                           TCGA-LGG   

                                              drs_uri    byte_size  \
0   drs://dg.4DFC:1b450a73-a16d-4de4-b140-92953350...      2850423   
1   drs://dg.4DFC:ef900de0-892c-4e2e-8eae-74caf55d...        15444   
2   drs://dg.4DFC:d43ef537-6cd0-446a-9bf0-fe0767ed...    356728009   
3   drs://dg.4DFC:58581feb-03c1-496c-9328-e8dda42c...     22617382   
4   drs://dg.4DFC:1fd6817e-fa11-4569-a740-86584eaf...      4543929   
..                                                ...          ...   
95  drs://dg.4DFC:70b44166-c646-43eb-8d76-b567bf93.

As you might expect, searching file gives us a huge number of results. This is great if you are surveying what kind of data is available, but is less useful for getting a coherent cohort. 

A better way to get files for a specific cohort is to chain your queries together, which we cover in the next tutorial [Chaining Queries](../ChainingQueries): Combine information from multiple endpoints, and build And/Or/Like and other advanced query strings.

Another useful way to look at high level information is to use our counts feature which returns summary information rather than the full search results. Check out the [Data Summaries tutorial](../DataSummaries) to try it.



---

<div style="background-color:#a2f2ed;color:black;padding:20px;">

<h3>File Field Definitions</h3>

<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>

    
<ul>
  <li><b>id:</b> The unique identifier for this file</li>
  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the file had there</li>
  <li><b>label:</b> The full name of the file</li>
  <li><b>data_catagory:</b> A desecription of the kind of general kind data the file holds</li>
  <li><b>data_type:</b> A more specific descripton of the data type</li>
  <li><b>file_format:</b> String to identify the full file extension including compression extensions</li>
  <li><b>associated_project:</b> The name the data center uses for the study this file was generated for</li>
  <li><b>drs_uri:</b> A unique identifier that can be used to retreive this specific file from a server</li>
  <li><b>byte_size:</b> Size of the file in bytes</li>
  <li><b>checksum:</b> The md5 value for the file</li>
  <li><b>data_modality:</b> Describes the biological nature of the information gathered as the result of an activity, independent of the technology or methods used to produce the information. Always one of "Genomic", "Proteomic", or "Imaging"</li>
  <li><b>imaging_modality:</b> For files with the `data_modality` of "Imaging", a descriptor for the image type</li>
  <li><b>dbgap_accession_number:</b> The project id number for this data on dbGaP</li>
</ul>  

</div>
    
---
